src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala

```diff
@@ -0,0 +1,101 @@
+package io.iohk.ethereum.vm
+
+import akka.util.ByteString
+import io.iohk.ethereum.crypto._
+import io.iohk.ethereum.domain.Address
+import GasFee.wordsForBytes
+import io.iohk.ethereum.utils.ByteUtils
+
+import scala.util.Try
+
+// scalastyle:off magic.number
+object PrecompiledContracts {
+
+  val EcDsaRecAddr = Address(1)
+  val Sha256Addr = Address(2)
+  val Rip160Addr = Address(3)
+  val IdAddr = Address(4)
+
+  val contracts = Map(
+    EcDsaRecAddr -> EllipticCurveRecovery,
+    Sha256Addr -> Sha256,
+    Rip160Addr -> Ripemp160,
+    IdAddr -> Identity
+  )
+
+  /**
+    * Given a [[ProgramContext]] it optionally runs a precompiled contract if the receiving address is one of
+    * the contracts
+    *
+    * @return None if the receiving address is not a precompiled contract
+    */
+  def runOptionally[W <: WorldStateProxy[W, S], S <: Storage[S]](context: ProgramContext[W, S]): Option[ProgramResult[W, S]] =
+    contracts.get(context.receivingAddr).map(_.run(context))
+
+  sealed trait PrecompiledContract {
+    protected def exec(inputData: ByteString): ByteString
+    protected def gas(inputData: ByteString): UInt256
+
+    def run[W <: WorldStateProxy[W, S], S <: Storage[S]](context: ProgramContext[W, S]): ProgramResult[W, S] = {
+      val g = gas(context.env.inputData)
+
+      val (result, error, gasRemaining): (ByteString, Option[ProgramError], UInt256) =
+        if (g < context.startGas)
```

Comment:
Should we allow the execution of the precompiled contract if `g==context.startGas`?